### PR TITLE
Read Tendermint's height & blocks from database

### PR DESCRIPTION
**Pull request by folex**

### CODE REVIEW

File: effects/tendermint-rpc/src/main/scala/fluence/effects/tendermint/rpc/websocket/TendermintWebsocketRpcImpl.scala [hunk 1]
```diff
@@ -55,6 +56,7 @@ class TendermintWebsocketRpcImpl[F[_]: ConcurrentEffect: Timer: Monad: ContextSh
   host: String,
   port: Int,
   httpRpc: TendermintHttpRpc[F],
+  blockstore: Blockstore[F],
```
Comment (alari) [inline effects/tendermint-rpc/src/main/scala/fluence/effects/tendermint/rpc/websocket/TendermintWebsocketRpcImpl.scala]: Please document why Blockstore is needed.

Before that, `tendermint-rpc` was really a remote procedure call library, but now it is integrated inside the blockstore. That's very surprising.

File: history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  /**
+   * Retrieves current height of the cluster from Tendermint's database
+   */
+  def getStorageHeight: EitherT[F, BlockstoreError, Long]
+}
+
+object Blockstore {
+  type RawKV[F[_]] = KVStore[F, Array[Byte], Array[Byte]]
+
+  val BlockStoreHeightKey: Array[Byte] = "blockStore".getBytes
```
Comment (alari) [inline history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala]: Docs please

File: history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  private def createSymlinks[F[_]: Log](
+    levelDbDir: Path
+  )(implicit F: Sync[F]) = {
+    import Files.{createSymbolicLink => createSymlink}
+
+    def ldbToSst(file: Path) = file.getFileName.toString.replaceFirst(".ldb$", ".sst")
```
Comment (alari) [inline history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala]: Docs please: it's unclear what problem is solved here until you already know it.

File: history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    )(p => F.delay(p.foreach(rmDir)).attempt.void)
+  }
+
+  private def rocksDbStore[F[_]: Log: Monad: LiftIO: ContextShift: Defer](p: Path): Resource[F, Blockstore[F]] =
+    RocksDBStore
+      .makeRaw[F](p.toString, createIfMissing = false, readOnly = true)
+      .map(kv => new BlockstoreImpl(kv))
+
+  // TODO: using MonadError here because caller (DockerWorkerServices) uses it, avoid doing that
+  private def raise[F[_]: Log: Sync, T, E <: Throwable](
```
Comment (alari) [inline history/tendermint-block-history/src/main/scala/fluence/effects/tendermint/block/history/db/Blockstore.scala]: Function's name make me expect that it always raises. But it doesn't.

File: effects/tendermint-rpc/src/main/scala/fluence/effects/tendermint/rpc/websocket/TendermintWebsocketRpcImpl.scala [hunk 5]
```diff
@@ -179,17 +194,33 @@ class TendermintWebsocketRpcImpl[F[_]: ConcurrentEffect: Timer: Monad: ContextSh
...
   ) = Traverse[List].sequence((from to to).map(loadBlock).toList)
 
-  private def getLastHeight: EitherT[F, EffectError, Long] = httpRpc.consensusHeight().leftMap(identity[EffectError])
+  private def getLastHeight(implicit log: Log[F]): EitherT[F, EffectError, Long] =
+    EitherT.liftF(traceBU("getLastHeight")).leftMap(identity[EffectError]) *>
+      blockstore.getStorageHeight.leftMap(identity[EffectError]).recoverWith {
+        case e =>
+          Log.eitherT[F, EffectError].warn(s"Error retrieving last height from blockstore", e) >>
+            httpRpc.consensusHeight().leftMap(identity[EffectError])
+      }
```
Comment (alari) [inline effects/tendermint-rpc/src/main/scala/fluence/effects/tendermint/rpc/websocket/TendermintWebsocketRpcImpl.scala]: Can you just make this `leftMap` once on the last line?

File: node/src/main/resources/docker/application.conf [hunk 6]
```diff
@@ -55,8 +54,10 @@ kademlia {
 
   join {
     seeds = [
-      "fluence://HYAXgtDyiuLzeGR4A7j5g8gYX4Poezt845PbanxeF32H:2YhKxiGRTm72MPxw1nr8v5yDyKRdtcYde5acM73fgnBzh655w9SYCniZDkNbwR3UXV76LBpvnNWDGuXqYRaE1BSE@207.154.210.151:25000", //node-stage-01
-      "fluence://Bx274w34MLanytDgxjnQ7a35eRGuao9XAZTC9KzYNzzy:5HDWMdpci62LGW1dbBZoXYVM176dktLtjUdvDY9yS4j261GZT3gMwgFBhUmdBc73WZGLiMPnH2uXqnUYQVLebFPF@46.101.221.231:25000", //node-stage-02
+      //node-stage-01
+      "fluence://DqxHrTprYzBDNqgzm1DXMNhop5DwK4EXNE8pj2rwfw4P:4wk5263LhXPfWDrLygKikQkk3d3cFYWvtxhtpEc6YVhjL53t5Ru5xtUqiawwQfQuwtido6NXgUQ9D3YrKjdNwcV6@46.101.151.125:25000",
```
Comment (alari) [inline node/src/main/resources/docker/application.conf]: Are you sure that this should be committed?

File: node/src/main/scala/fluence/node/MakeResource.scala [hunk 7]
```diff
@@ -122,4 +128,23 @@ object MakeResource {
       )
       _ ← fiberDef.complete(fiber)
     } yield ()
+
+  /**
+   * Allocates resource in (possible concurrent) context, defined by `evalOn` function
+   * Use this when you need to go from Resource[F, T] to Resource[ F, F[T] ] (especially concurrently)
+   *
+   * @param resource Resource to be allocated
+   * @param evalOn Description of how to evaluate resource
```
Comment (alari) [inline node/src/main/scala/fluence/node/MakeResource.scala]: It doesn't explain anything, could you rephrase? What could be an example of such a function? What should it do? How it describes the resource evaluation? What does "resource evaluation" mean?

File: node/src/main/resources/docker/application.conf (hunk 6 above)
Comment (folex) [reply node/src/main/resources/docker/application.conf]: Well, it updates stage node addresses, but maybe it is better to put devnet nodes here. I'll add devnet nodes.
